### PR TITLE
Search result type tabs

### PR DESCRIPTION
**Pull request by attfarhan**
image

Adds search result type tabs to the results page.

Not included in this PR: 
- Error message when invalid filters are used in a query. (E.g. if `after:` is used in a code or symbol search)

Fixes https://github.com/sourcegraph/sourcegraph/issues/5378.

### CODE REVIEW

**Comment by attfarhan** [comment]
@christinaforney could you take a look at the GIF/pull down this branch and see if the design/UI is OK? I mimicked the wireframe in the RFC but wanted to confirm this is along the lines of what you were thinking.

**Comment by christinaforney** [comment]
This looks great! I don't like the jumpiness due to the shifting suggested filters, but that should go away with some of the upcoming changes. If you can make it so it holds space for one line of filters (since there is always at least one), that would reduce the jitter by quite a bit, but only if it isn't too much effort.

I assume these are the tab styles we have by default for the app? Or did you create these styles yourself?

**Comment by attfarhan** [comment]
I created the styles, I don't think we have any tabs in the app other than those in the settings area, which is quite different: 

image

**Comment by christinaforney** [comment]
For consistency, we should use the same styles for tabs throughout the site. Can you update the new tabs to look like the ones from settings please?

File: web/src/search/results/SearchResultTab.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+const tabIsActive = (builtQuery: string, location: H.Location): boolean => location.search === '?' + builtQuery
+
+const tabIsActiveTrue = (): boolean => true
+const tabIsActiveFalse = (): boolean => false
+
+export const SearchResultTab: React.FunctionComponent<Props> = props => {
+    const q = toggleSearchType(props.query, props.type)
+    const newURLSearchParam = buildSearchURLQuery(q)
+
+    const isActiveFunc = tabIsActive(newURLSearchParam, props.location) ? tabIsActiveTrue : tabIsActiveFalse
```
Comment (lguychard) [inline web/src/search/results/SearchResultTab.tsx]: I suggest using lodash's `constant` for this:

```suggestion
    const isActiveFunc = constant(location.search === `?${builtQuery}`)
```

File: web/src/e2e/e2e.test.ts [hunk 2]
```diff
@@ -100,1012 +100,1055 @@ describe('e2e test suite', () => {
...
-
-    describe('Visual tests', () => {
-        test('Repositories list', async () => {
-            await driver.page.goto(sourcegraphBaseUrl + '/site-admin/repositories?query=gorilla%2Fmux')
-            await driver.page.waitForSelector('a[href="/github.com/gorilla/mux"]', { visible: true })
-            await percySnapshot(driver.page, 'Repositories list')
-        })
-
-        test('Search results repo', async () => {
+    // describe('Core functionality', () => {
```
Comment (lguychard) [inline web/src/e2e/e2e.test.ts]: Could you remove these comments so that the diff displays correctly?

File: web/src/e2e/e2e.test.ts [hunk 3]
```diff
@@ -100,1012 +100,1055 @@ describe('e2e test suite', () => {
...
-                await driver.page.evaluate(() =>
-                    Array.from(document.querySelector('.theme')!.classList).filter(c => c.startsWith('theme-'))
-                )
-            ).toEqual(['theme-light'])
-        })
-    })
+            expect(tabs.length).toEqual(5)
 
-    describe('Repository component', () => {
-        const blobTableSelector = '.e2e-blob > ta
```
Comment (lguychard) [inline web/src/e2e/e2e.test.ts]:
```suggestion
            expect(tabs).toBe(['Code', 'Diffs', 'Symbols', 'Repos'])
```

File: web/src/search/helpers.tsx [hunk 4]
```diff
@@ -74,6 +75,52 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
     return (query.substring(0, idx).trim() + ' ' + query.substring(idx + searchFilter.length).trim()).trim()
 }
 
+export function getSearchTypeFromQuery(query: string): SEARCH_TYPES {
+    // RegExp to match `type:$TYPE` in any part of a query.
+    const matchSearchType = /(\b|^)type:\w*(\s*|$)/
+    const getTypeName = /(\b|^)type:(?<type>\w*)(\s|$)/
+    // const m = query.match(getTypeName)
+    if (matchSearchType.test(query)) {
```
Comment (lguychard) [inline web/src/search/helpers.tsx]: Do you really need both checks? `query.match(getTypeName)` will return `null` if there's no match.

File: web/src/search/helpers.tsx [hunk 5]
```diff
@@ -74,6 +75,52 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
...
+        }
+    }
+
+    return ''
+}
+
+export function toggleSearchType(query: string, searchType: SEARCH_TYPES): string {
+    if (searchType === '') {
+        const replaceSearchType = /(\b|^)type:\w*(\s|$)/
+        // RegExp to match `repogroup:sample` in any part of a query.
```
Comment (lguychard) [inline web/src/search/helpers.tsx]: This comment appears to be unrelated?

File: web/src/search/helpers.tsx [hunk 6]
```diff
@@ -74,6 +75,52 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
...
+        const matches = query.match(getTypeName)
+        if (matches && matches.groups && matches.groups.type) {
+            return matches.groups.type as SEARCH_TYPES
+        }
+    }
+
+    return ''
+}
+
+export function toggleSearchType(query: string, searchType: SEARCH_TYPES): string {
```
Comment (lguychard) [inline web/src/search/helpers.tsx]: This function and `getSearchTypeFromQuery()` would benefit from being unit-tested.

I think it could be simplified quite a bit as well (I didn't try to run this and it should also be unit-tested 🙂):

```typescript
export function toggleSearchType(query: string, searchType: SEARCH_TYPES): string {
    const match = query.match(/(?:\b|^)(type:\w*)(\s|$)/)
    if (!match) {
        // Add searchType filter if needed
        return searchType ? `${query} type:${searchType}` : query
    }
    if (match[1] === `type${searchType}`) {
        // Query already contains the right searchType
        return query
    }
    // Replace existing searchType in query
    return query.replace(match[1], searchType ? `type:${searchType}` : '')
}
```

File: web/src/search/results/SearchResultTab.tsx [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    symbol: 'Symbols',
+    repo: 'Repos',
+}
+
+const tabIsActive = (builtQuery: string, location: H.Location): boolean => location.search === '?' + builtQuery
+
+const tabIsActiveTrue = (): boolean => true
+const tabIsActiveFalse = (): boolean => false
+
+export const SearchResultTab: React.FunctionComponent<Props> = props => {
```
Comment (lguychard) [inline web/src/search/results/SearchResultTab.tsx]: Should this be named `SearchResultTabHeader` since it doesn't actually contain the tab content?

File: web/src/search/helpers.tsx [hunk 8]
```diff
@@ -74,6 +75,52 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
...
 }
 
+export function getSearchTypeFromQuery(query: string): SEARCH_TYPES {
+    // RegExp to match `type:$TYPE` in any part of a query.
+    const matchSearchType = /(\b|^)type:\w*(\s*|$)/
+    const getTypeName = /(\b|^)type:(?<type>\w*)(\s|$)/
+    // const m = query.match(getTypeName)
+    if (matchSearchType.test(query)) {
+        const matches = query.match(getTypeName)
+        if (matches && matches.groups && matches.groups.type) {
```
Comment (lguychard) [inline web/src/search/helpers.tsx]: Should this perform validation before casting the value as a valid search type? What is the expected behaviour if the user types `type:commmit` in his query by mistake? Will the UI gracefully fallback to displaying the code search results?

File: web/src/search/helpers.tsx [hunk 9]
```diff
@@ -74,6 +75,52 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
...
+
+        if (matchSearchType.test(query)) {
+            query = query.replace(replaceSearchType, '')
+        }
+
+        return query
+    }
+
+    const idx = queryIndexOfScope(query, 'type:' + searchType)
+    if (idx >= 0) {
```
Comment (felixfbecker) [inline web/src/search/helpers.tsx]: Can you explain this check in a comment?

File: web/src/search/helpers.tsx [hunk 10]
```diff
@@ -74,6 +75,52 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
...
+    const getTypeName = /(\b|^)type:(?<type>\w*)(\s|$)/
+    // const m = query.match(getTypeName)
+    if (matchSearchType.test(query)) {
+        const matches = query.match(getTypeName)
+        if (matches && matches.groups && matches.groups.type) {
+            return matches.groups.type as SEARCH_TYPES
+        }
+    }
+
+    return ''
```
Comment (felixfbecker) [inline web/src/search/helpers.tsx]: Prefer `undefined` to represent "no search type"

File: web/src/search/helpers.tsx [hunk 11]
```diff
@@ -74,6 +75,52 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
     return (query.substring(0, idx).trim() + ' ' + query.substring(idx + searchFilter.length).trim()).trim()
 }
 
+export function getSearchTypeFromQuery(query: string): SEARCH_TYPES {
+    // RegExp to match `type:$TYPE` in any part of a query.
+    const matchSearchType = /(\b|^)type:\w*(\s*|$)/
+    const getTypeName = /(\b|^)type:(?<type>\w*)(\s|$)/
```
Comment (felixfbecker) [inline web/src/search/helpers.tsx]: `\b` includes the start of the string
```suggestion
    const getTypeName = /\btype:(?<type>\w*)(\s|$)/
```

File: web/src/search/helpers.tsx (hunk 8 above)
Comment (attfarhan) [reply web/src/search/helpers.tsx]: Can you clarify what you mean here? If the user types in `type:commit` in his query, the "Commits" tab will be shown as active. Why would it then show code search results?

Comment (lguychard) [reply web/src/search/helpers.tsx]: Note the typo in `commit` - sorry I should've made it more obvious. Currently this code doesn't check that the type provided by the user is one of the valid search types. It's also unclear to me what happens if the user types in an invalid search type.

Comment (attfarhan) [reply web/src/search/helpers.tsx]: Oh got it. The search that occurs actually will return "No results" when there is an invalid `type` field. Since the `type` field doesn't match one of the valid search types, the UI will show none of the tabs as active. (The tabs don't actually determine what results are shown btw, it just shows the active tab based on the `type` fields in the query)

File: web/src/e2e/e2e.test.ts [hunk 12]
```diff
@@ -1109,4 +1109,42 @@ describe('e2e test suite', () => {
...
+            expect(activeTab).toEqual(1)
+
+            const label = await driver.page.evaluate(
+                () => document.querySelector('.e2e-search-result-tab--active')!.textContent || ''
+            )
+            expect(label).toEqual('Code')
+        })
+
+        test('Clicking search results tabs updates query and URL', async () => {
+            await driver.page.click('.e2e-search-result-tab-diff')
```
Comment (lguychard) [inline web/src/e2e/e2e.test.ts]: What would you think of DRYing this test?

```typescript
for (const searchType of ['diff', 'commit', 'symbol', 'repo']) {
    await driver.page.click(`.e2e-search-result-tab-${searchType}`)
    await driver.assertWindowLocation(`/search?q=repo:%5Egithub.com/gorilla/mux%24+type:${searchType}`)
}
```

File: web/src/search/helpers.test.tsx [hunk 13]
```diff
@@ -13,4 +15,49 @@ describe('search/helpers', () => {
             /* noop */
         })
     })
+
+    describe('getSearchTypeFromQuery()', () => {
```
Comment (lguychard) [inline web/src/search/helpers.test.tsx]: Typically, best practices would recommend not to write tests directly in a `describe()` block - the describe should group specific `test()` together, each covering a given behaviour, eg.:

```typescript
describe('getSearchTypeFromQuery()', () => {

    test('parses the searchType in simple queries', () => {
        expect(getSearchTypeFromQuery('type:diff')).toEqual('diff')
        expect(getSearchTypeFromQuery('type:commit')).toEqual('commit')
        expect(getSearchTypeFromQuery('type:symbol')).toEqual('symbol')
        expect(getSearchTypeFromQuery('type:repo')).toEqual('repo')
    })

    test('returns null when the query does not contain a searchType', () => {
        expect(getSearchTypeFromQuery('code')).toEqual(null)
    })

    test('parses the searchType in complex queries', () => {
        expect(getSearchTypeFromQuery('test type:diff')).toEqual('diff')
        expect(getSearchTypeFromQuery('type:diff test')).toEqual('diff')
        expect(getSearchTypeFromQuery('repo:^github.com/sourcegraph/sourcegraph type:diff test')).toEqual('diff')
        expect(getSearchTypeFromQuery('type:diff repo:^github.com/sourcegraph/sourcegraph test')).toEqual('diff')
        expect(getSearchTypeFromQuery('type:diff type:commit repo:^github.com/sourcegraph/sourcegraph test')).toEqual(
            'diff'
        )
    })

    test('returns symbol when multiple searchTypes, including symbol, are specified', () => {
        /** Edge case. If there are multiple type filters and `type:symbol` is one of them, symbol results always get returned. */
        expect(getSearchTypeFromQuery('type:diff type:symbol repo:^github.com/sourcegraph/sourcegraph test')).toEqual(
            'symbol'
        )
    })

})
```

File: web/src/search/helpers.tsx [hunk 14]
```diff
@@ -74,6 +75,50 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
...
+    const matches = query.match(getTypeName)
+
+    if (matches && matches.groups && matches.groups.type) {
+        /**
+         * In an edge case where multiple `type:` filters are used, if
+         * `type:symbol` is included, symbol results be returned, regardless of order,
+         * so we must check for `type:symbol`. For other types,
+         * the first `type` filter appearing in the query is applied.
+         **/
+        const getSymbolType = /\btype:symbol\b/
```
Comment (lguychard) [inline web/src/search/helpers.tsx]:
```suggestion
        const symbolTypeRegex = /\btype:symbol\b/
```

`getSymbolType` sounds like a function name (same for `getTypeName` above)

File: web/src/search/helpers.test.tsx [hunk 15]
```diff
@@ -13,4 +15,49 @@ describe('search/helpers', () => {
...
+        expect(getSearchTypeFromQuery('type:diff type:commit repo:^github.com/sourcegraph/sourcegraph test')).toEqual(
+            'diff'
+        )
+        /** Edge case. If there are multiple type filters and `type:symbol` is one of them, symbol results always get returned. */
+        expect(getSearchTypeFromQuery('type:diff type:symbol repo:^github.com/sourcegraph/sourcegraph test')).toEqual(
+            'symbol'
+        )
+    })
+
+    describe('toggleSearchType()', () => {
```
Comment (lguychard) [inline web/src/search/helpers.test.tsx]: The `toggleSearchType()` tests should probably include a corresponding test for the `type:symbol` edge case.

File: web/src/search/helpers.tsx [hunk 16]
```diff
@@ -74,6 +75,50 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
...
 
+export function getSearchTypeFromQuery(query: string): SearchType {
+    /** RegExp to match `type:$TYPE` in any part of a query. */
+    const getTypeName = /\btype:(?<type>\w*)\b/
+    const matches = query.match(getTypeName)
+
+    if (matches && matches.groups && matches.groups.type) {
+        /**
+         * In an edge case where multiple `type:` filters are used, if
+         * `type:symbol` is included, symbol results be returned, regardless of order,
```
Comment (lguychard) [inline web/src/search/helpers.tsx]: What should happen when multiple types are included and `symbol` is **not** one of them?

File: web/src/search/helpers.tsx [hunk 17]
```diff
@@ -74,6 +75,50 @@ export function toggleSearchFilter(query: string, searchFilter: string): string
     return (query.substring(0, idx).trim() + ' ' + query.substring(idx + searchFilter.length).trim()).trim()
 }
 
+export function getSearchTypeFromQuery(query: string): SearchType {
+    /** RegExp to match `type:$TYPE` in any part of a query. */
+    const getTypeName = /\btype:(?<type>\w*)\b/
```
Comment (lguychard) [inline web/src/search/helpers.tsx]: Should this only specify valid searchTypes?

```suggestion
    const getTypeName = /\btype:(?<type>diff|commit|repo|symbol)\b/
```

File: web/src/search/helpers.test.tsx [hunk 18]
```diff
@@ -13,4 +15,49 @@ describe('search/helpers', () => {
...
+        expect(getSearchTypeFromQuery('code')).toEqual(null)
+
+        expect(getSearchTypeFromQuery('test type:diff')).toEqual('diff')
+        expect(getSearchTypeFromQuery('type:diff test')).toEqual('diff')
+        expect(getSearchTypeFromQuery('repo:^github.com/sourcegraph/sourcegraph type:diff test')).toEqual('diff')
+        expect(getSearchTypeFromQuery('type:diff repo:^github.com/sourcegraph/sourcegraph test')).toEqual('diff')
+        expect(getSearchTypeFromQuery('type:diff type:commit repo:^github.com/sourcegraph/sourcegraph test')).toEqual(
+            'diff'
+        )
+        /** Edge case. If there are multiple type filters and `type:symbol` is one of them, symbol results always get returned. */
```
Comment (lguychard) [inline web/src/search/helpers.test.tsx]: As a clarification for [my previous comment](https://github.com/sourcegraph/sourcegraph/pull/5402#discussion_r320742058) (apologies if it wasn't clear enough!):
- If a comment should be a label for a constant, function, variable, type, etc... Basically, if it would be helpful when displayed in a hover tooltip, use valid JSDoc. In the linked comment, it was helpful to have a JSDoc docblock for `SearchType`.
- Otherwise, for single-line comments like this, `//` is fine.

File: web/src/search/helpers.tsx (hunk 16 above)
Comment (attfarhan) [reply web/src/search/helpers.tsx]: The first `type` filter wins, as specified in this same comment a couple lines down.

File: web/src/e2e/e2e.test.ts (hunk 12 above)
Comment (attfarhan) [reply web/src/e2e/e2e.test.ts]: Yeah we should do this, thanks

File: web/src/search/results/SearchResultTab.tsx [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+    commit: 'Commits',
+    symbol: 'Symbols',
+    repo: 'Repos',
+}
+
+export const SearchResultTabHeader: React.FunctionComponent<Props> = props => {
+    const q = toggleSearchType(props.query, props.type)
+    const builtURLQuery = buildSearchURLQuery(q)
+
+    const isActiveFunc = constant(location.search === `?${builtURLQuery}`)
```
Comment (lguychard) [inline web/src/search/results/SearchResultTab.tsx]: Looks like you're accessing the global `location` instead of the one injected through `props`? It also looks like `props.history` is not needed.

You could destructure `props` and avoid declaring a separate `type` variable:

```typescript
export const SearchResultTabHeader: React.FunctionComponent<Props> = ({ location, type, query }) => {
```
